### PR TITLE
feat(server): Support limiting the number of open connections.

### LINKED ISSUE #1496: introduce maxclients flag
DragonflyDB should be able to protect itself from opening too many connections on the primary port. Otherwise, it will risk reaching max number of open files. 

Dragonfly can support a high number by default, I would say 64k is ok as a default for the flag.
Redis sets it as 1000: https://github.com/redis/redis/blob/6bf9b144ef722fb301c065125a6bf6cd385841bf/src/config.c#L3188 

Also, we should provide a friendly warning on startup if max files ulimit is lower than `maxclients` - see https://github.com/redis/redis/blob/6bf9b144ef722fb301c065125a6bf6cd385841bf/src/server.c#L2278 and 
https://github.com/redis/redis/blob/6bf9b144ef722fb301c065125a6bf6cd385841bf/src/config.c#L2539 for example

### PR DESCRIPTION
**Pull request by royjacobson**
Close #1496.

Also, make `CONFIG GET` use config_registry so it's easier to add flags.

### CODE REVIEW

File: src/facade/dragonfly_listener.cc [hunk 1]
```diff
@@ -267,6 +267,10 @@ void Listener::OnConnectionClose(util::Connection* conn) {
   }
 }
 
+void Listener::OnMaxConnectionsReached(util::FiberSocketBase* sock) {
```
Comment (chakaz) [inline src/facade/dragonfly_listener.cc:270]: What calls this? It looks like nothing(?), so maybe move it to a different PR?

Comment (romange) [reply src/facade/dragonfly_listener.cc:270]: helio framework calls this. it's a virtual hook

File: src/server/server_family.cc [hunk 2]
```diff
@@ -569,6 +571,21 @@ void ServerFamily::Init(util::AcceptServer* acceptor, std::vector<facade::Listen
...
 
+  for (auto* listener : listeners_) {
+    listener->SetMaxClients(absl::GetFlag(FLAGS_maxclients));
+  }
+  config_registry.Register("maxclients", [this](const absl::CommandLineFlag& flag) {
+    auto res = flag.TryGet<uint32_t>();
+    if (!res)
+      return false;
+
+    for (auto* listener : listeners_) {
```
[the remark this answers is not in the document]
Comment (royjacobson) [reply src/server/server_family.cc]: Changed (and added a test accordingly).

I added `IsAdmin/SetAdmin` methods to `Listener`, but I don't think it's an overkill, we already wanted something similar when we spoke about hiding admin connections from `CLIENT LIST`.

File: src/facade/dragonfly_listener.h [hunk 3]
```diff
@@ -35,12 +35,16 @@ class Listener : public util::ListenerInterface {
   bool AwaitDispatches(absl::Duration timeout,
                        const std::function<bool(util::Connection*)>& filter);
 
+  bool IsAdminInterface() const;
+  void SetAdminInterface(bool is_admin = true);
```
Comment (chakaz) [inline src/facade/dragonfly_listener.h:39]: We already have Connection::IsAdmin(), could we maybe consolidate them to use the same logic? (either inspect the port or explicitly set like here)

**Review by chakaz — DISMISSED** [review]

File: src/server/config_registry.cc [hunk 4]
```diff
@@ -39,6 +39,19 @@ bool ConfigRegistry::Set(std::string_view config_name, std::string_view value) {
   return cb(*flag);
 }
 
+std::optional<std::string> ConfigRegistry::Get(std::string_view config_name) {
+  unique_lock lk(mu_);
+  auto it = registry_.find(config_name);
+  if (it == registry_.end())
```
Comment (kostasrim) [inline src/server/config_registry.cc:45]: You can use `contains`

File: src/server/config_registry.cc [hunk 5]
```diff
@@ -39,6 +39,19 @@ bool ConfigRegistry::Set(std::string_view config_name, std::string_view value) {
...
+    return std::nullopt;
+  auto cb = it->second;
```
Comment (kostasrim) [inline src/server/config_registry.cc:47]: dead code, `cb` not used anywhere below

File: tests/dragonfly/config_test.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+        with pytest.raises(redis.exceptions.ConnectionError):
+            async with server.client() as client2:
+                await client2.get("test")
+
+        # Check that admin connections are not limited.
+        async with RedisClient(port=server.admin_port) as admin_client:
+            await admin_client.get("test")
+
+        await client1.execute_command("CONFIG SET maxclients 3")
```
Comment (kostasrim) [inline tests/dragonfly/config_test.py:24]: I was thinking of adding a `get maxclients` here but it's not really needed since you implicitly test the limit on the test case below...

**Review by kostasrim — COMMENTED** [review]
🧑‍🍳 🧑‍🍳 🧑‍🍳 nice work!

File: src/facade/dragonfly_listener.h (hunk 3 above)
Comment (royjacobson) [reply src/facade/dragonfly_listener.h:39]: Oooo, I did NOT expect the implementation of `Connection::IsAdmin` to be what it is :laughing:
Changed it to query the Listener.

File: src/facade/dragonfly_listener.cc [hunk 7]
```diff
@@ -267,6 +275,10 @@ void Listener::OnConnectionClose(util::Connection* conn) {
...
+  sock->Write(io::Buffer("-ERR max number of clients reached\r\n"));
```
Comment (chakaz) [inline src/facade/dragonfly_listener.cc:279]: How will this be handled by the client?
I _imagine_ that it will receive an error before even sending a command, so will it even know how to treat it?
Also, don't we want to close the connection? (or is it implicitly done by the framework?)

File: src/facade/dragonfly_connection.cc [hunk 8]
```diff
@@ -399,8 +400,7 @@ uint32_t Connection::GetClientId() const {
 }
 
 bool Connection::IsAdmin() const {
-  uint16_t admin_port = absl::GetFlag(FLAGS_admin_port);
-  return socket_->LocalEndpoint().port() == admin_port;
+  return static_cast<Listener*>(owner())->IsAdminInterface();
```
Comment (chakaz) [inline src/facade/dragonfly_connection.cc:403]: Can we add `IsAdminInterface()` API to `ListenerInterface`?
I realize it's perhaps not clean, but having a downcast is messy and can lead to issues down the road

Comment (romange) [reply src/facade/dragonfly_connection.cc:403]: `IsAdminInterface` is dragonfly specific. On the other side, facade::Connection is always issued by facade::Listener.

Comment (chakaz) [reply src/facade/dragonfly_connection.cc:403]: I'm open to any suggestions of removing hardcoded downcasts :)
We could even template-ize Listener to have facade Listener return its own connection

Comment (romange) [reply src/facade/dragonfly_connection.cc:403]: My **personal** opinion is that we should avoid casts if there is a reasonable design choice that allows it. And if not - we can have them and not feel bad about it. Specifically, any virtual inheritance implies the possibility of downcasts and it's fine.

Comment (royjacobson) [reply src/facade/dragonfly_connection.cc:403]: I don't like the cast either, if we start doing it too much I think we should find out a better way, but for now this was the shortest way I found to do this :)

File: src/facade/dragonfly_listener.cc (hunk 7 above)
Comment (royjacobson) [reply src/facade/dragonfly_listener.cc:279]: The clients I tested (python, redis-cli) seemed to handle it fine. redis does behave a bit differently (the socket doesn't close immediately and the message is delayed until the first command).

But then I tried to DDoS redis by opening a bunch of connections and leaving them open, and apparently it will start doing what we do.

and yeah, the framework is closing the connection :)

**Review by chakaz — APPROVED** [review]
